result-watcher.bal

```diff
@@ -300,79 +365,32 @@ public isolated class ResultWatcher {
-                        where tempSubsteps.indexOf(substep) == ()
-                        select substep;
-                    // add in db
-                    foreach var substep in newSubsteps {
-                        check database:updateTimelineSubstep(self.stepId, substep); // TODO: perhaps nicer to pass TimelineSubstep
-                    }
-
-                    // TODO: check that UI didn't delete substeps
-                    // TODO: neuer substep -> alle davor auf cleared setzen und optional warning/error
-                    // TODO: checken ob welche rausfliegen oder daziwschen welche eingefügt werden... -> error, hrefs vergleichen, substep
-                    // wenn substepid nicht gesetzt darf auch nicht nachträglich gesetzt werden... 
-
-                    // write progress into db - 
-                    if progress != () && newSubsteps.length() > 0 {
-                        check database:updateTimelineProgress(self.stepId, progress);
-                    }
-
-                    // save current task log
-                    if taskLog != () {
-                        check database:updateTimelineTaskLog(self.stepId, taskLog);
-                    }
-                    check commit;
-                } on fail error err {
-                    io:println(err);
-                }
-
-                do {
+            do {
+                ResultProcessor processor = new (result, self.experimentId, self.stepId, self.resultEndpoint);
+                boolean isChanged = check processor.processIntermediateResult();
+                if isChanged {
                     check self.unschedule();
-                    // TODO: Better way to reschedule this???
+                    // TODO: Probably needs to be changed in the future
                     (decimal|int)[] initialIntervals = [2, 10, 5, 10, 10, 60, 30, 20, 60, 10, 600];
                     check self.schedule(...initialIntervals);
-                } on fail error e {
-                    lock {
-                        self.errorCounter += 1;
-                    }
-                    io:println(e);
                 }
-            } else {
-                // TODO: update progress?
-                return; // nothing to do, still waiting for result
+            } on fail error e {
+                lock {
+                    self.errorCounter += 1;
+                }
+                io:println(e);
             }
-        }
-        do {
-            ResultProcessor processor = new (result, self.experimentId, self.stepId, self.resultEndpoint);
-            check processor.processResult();
-            check self.unschedule();
-        } on fail error e {
-            lock {
-                self.errorCounter += 1;
+        } else {
+            do {
+                ResultProcessor processor = new (result, self.experimentId, self.stepId, self.resultEndpoint);
+                check processor.processResult();
+                check self.unschedule();
```

Comment:
This is the point where the result watcher should remove itself from the registry. The result processor should not have to worry about the job unscheduling.

Reply:
That means we should also move the removeResultWatcher calls in saveSuccessfulResult and saveErrorResult to ResultWatcher?

Reply:
That seams like a good idea for now. Then at least we have alll the scheduling and unscheduling logic in the watcher and not scattered around as much. Though you can decide to leave this change for later.